doc/rst/developer/bestPractices/MemoryProfiling.rst

```diff
@@ -0,0 +1,85 @@
+.. _best-practices-memory-profiling:
+
+Memory Profiling
+================
+
+When using a non-standard memory allocator such as jemalloc, the allocator may
+provide profiling capabilities that can be useful for understanding memory
+usage patterns. For Chapel programs, this can be used to augment the existing
+Chapel runtime's memory tracking capabilities. For the Chapel compiler, this
+can be used to understand the memory usage of the compiler itself.
+
+Profiling the compiler with jemalloc
+------------------------------------
+
+To profile the compiler with jemalloc, make sure you build the compiler
+with the following environment variables set:
+
+.. code-block:: bash
+
+   export CHPL_HOST_MEM=jemalloc
+   export CHPL_HOST_JEMALLOC="bundled"
+   export CHPL_JEMALLOC_ENABLE_PROFILING=1
+
+Then the compiler can be run as follows:
+
+.. code-block:: bash
+
+   MALLOC_CONF='prof:true,prof_prefix:jeprof.out,prof_final:true' \\
+     chpl myProgram.chpl
+
+The value of ``prof_prefix`` is the prefix for the output files.
+
+It may be useful to pass ``--no-compiler-driver`` to the compiler. Without this
+flag, the compiler functions as a driver and spawns several subprocesses for
+different phases of the compiler. This will results in multiple profiling
+output files, one for each subprocess. If you want to profile the compiler as a
+single process, use ``--no-compiler-driver``.
+
+Output files can be interpreted using ``jeprof`` (which is built with
+jemalloc). For example, the following will generate a simple text report,
+ordered by cumulative memory usage:
+
+.. code-block:: bash
+
+   JEMALLOC_SUBDIR=$(./util/chplenv/chpl_jemalloc.py --host --cfg-path)
+   ./third-party/jemalloc/install/$JEMALLOC_SUBDIR/bin/jeprof \\
+     --text --cum $(which chpl) path/to/profile/data/jeprof.out
+
+
+Profiling Chapel programs with jemalloc
+---------------------------------------
+
+.. warning::
+
+   When writing these docs, I was unable to get jemalloc's profiling to give
+   good results for Chapel programs. I am including these docs here for
```

Comment:
I'd be more explicit about what "good results" means here.  Did you get bad performance?  Was the information it printed incorrect in some way?

I'm also debating whether a date/release number would be good to include here.  It seems like it'd be nice to improve but have no idea when we'd prioritize that so maybe an old date would help encourage people to ask if there's been effort since then? (Which would in turn encourage us to prioritize it)

Reply:
By "good results", I mean I was unable to get any results at all. jemalloc would report 0 allocations, which is just plain wrong. I suspect there is more work that needs to be done to integrate with the arenas the runtime uses.

The goal of this was to profile the compiler, so I am not that invested in making it work with chapel programs yet (but still wanted what I had done to be recorded)

Reply:
I just added a caveat to the docs about this

Reply:
No, we mostly do dates for internal comments.  Release number definitely has more precedent in user-facing documentation